app/post_acknowledgements.go

```diff
@@ -0,0 +1,141 @@
+		}
+	}
+
+	a.Srv().Go(func() {
+		a.sendAcknowledgementEvent(model.WebsocketEventAcknowledgementAdded, acknowledgement, post)
+	})
+
+	return acknowledgement, nil
+}
+
+func (a *App) DeleteAcknowledgementForPost(c *request.Context, userID, postID string) *model.AppError {
+	post, err := a.GetSinglePost(postID, false)
+	if err != nil {
+		return err
+	}
+
+	channel, err := a.GetChannel(c, post.ChannelId)
+	if err != nil {
+		return err
+	}
+
+	if channel.DeleteAt > 0 {
+		return model.NewAppError("DeleteAcknowledgementForPost", "api.acknowledgement.delete.archived_channel.app_error", nil, "", http.StatusForbidden)
+	}
+
+	oldAck, nErr := a.Srv().Store().PostAcknowledgement().Get(userID, postID)
+
+	if nErr != nil {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(nErr, &nfErr):
+			return model.NewAppError("GetPostAcknowledgement", "app.acknowledgement.get.app_error", nil, "", http.StatusNotFound).Wrap(nErr)
+		default:
+			return model.NewAppError("GetPostAcknowledgement", "app.acknowledgement.get.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+		}
+	}
+
+	if model.GetMillis()-oldAck.AcknowledgedAt > 5*60*1000 {
+		return model.NewAppError("DeleteAcknowledgementForPost", "api.acknowledgement.delete.deadline.app_error", nil, "", http.StatusForbidden)
+	}
+
+	nErr = a.Srv().Store().PostAcknowledgement().Delete(oldAck)
+
+	if nErr != nil {
+		return model.NewAppError("DeleteAcknowledgementForPost", "app.acknowledgement.delete_all_with_emoji_name.get_acknowledgement.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+	}
+
+	oldAck.AcknowledgedAt = 0
+
+	a.Srv().Go(func() {
+		a.sendAcknowledgementEvent(model.WebsocketEventAcknowledgementRemoved, oldAck, post)
+	})
+
+	return nil
+}
+
+func (a *App) GetAcknowledgementsForPost(postID string) ([]*model.PostAcknowledgement, *model.AppError) {
+	_, err := a.GetSinglePost(postID, false)
+	if err != nil {
+		return nil, err
+	}
+
+	acknowledgements, nErr := a.Srv().Store().PostAcknowledgement().GetForPost(postID)
+	if nErr != nil {
+		return nil, model.NewAppError("GetAcknowledgementsForPost", "app.acknowledgement.getforpost.get.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+	}
+
+	return acknowledgements, nil
+}
+
+func (a *App) GetAcknowledgementsForPostList(postList *model.PostList) (map[string][]*model.PostAcknowledgement, *model.AppError) {
+	acknowledgements, err := a.Srv().Store().PostAcknowledgement().GetForPosts(postList.Order)
+
+	if err != nil {
+		return nil, model.NewAppError("GetPostAcknowledgementsForPostList", "app.acknowledgement.get.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	acknowledgementsMap := make(map[string][]*model.PostAcknowledgement)
+
+	for _, ack := range acknowledgements {
+		if _, ok := acknowledgementsMap[ack.PostId]; !ok {
+			acknowledgementsMap[ack.PostId] = make([]*model.PostAcknowledgement, 0)
+		}
```

Comment:
Can we eliminate this check and just rely on the `nil` handling semantics of `append`?

Reply:
I'm not sure how to do that. Don't we need to initialize the slice for each post?

Reply:
Assuming I'm reading the code correctly (and I've already done that wrong once so far ;P), `acknowledgementsMap[ack.PostId]` starts out `nil`. When given a nil value (aka the empty value for a slice, itself subtly different than an empty slice), `append` simply allocates a new slice and returns that, making it safe, for example, to do this:

```go
var list []string
list = append(list, "first")
list = append(list, "second")
```

So the code in this case could just read:
```go
for _, ack := range acknowledgements {
    acknowledgementsMap[ack.PostId] = append(acknowledgementsMap[ack.PostId], ack)
}
```

Note that the map itself has to be initialized: the `nil` value for a map is safe to read from, but not safe to write to.

Reply:
oh I see, didn't know about that, thanks!